InfoHandler.cpp

```diff
@@ -83,57 +104,41 @@ void InfoHandler::applyMethodType(BinaryViewRef bv, const ObjectiveNinja::ClassI
         return;
     }
 
-    // Shorthand for formatting an individual "part" of the type signature.
-    auto partForIndex = [selectorTokens, typeTokens](size_t i) {
-        std::string argName;
-
-        // Indices 0, 1, and 2 are the function return type, self parameter, and
-        // selector parameter, respectively. Indices 3+ are the actual
-        // arguments to the function.
-        if (i == 0)
-            argName = "";
-        else if (i == 1)
-            argName = "self";
-        else if (i == 2)
-            argName = "sel";
-        else if (i - 3 < selectorTokens.size())
-            argName = selectorTokens[i - 3];
-
-        return typeTokens[i] + " " + argName;
-    };
-
-    // Build the type string for the method.
-    std::string typeString;
-    for (size_t i = 0; i < typeTokens.size(); ++i) {
-        std::string suffix;
-        auto part = partForIndex(i);
-
-        // The underscore being used as the function name here is critically
-        // important as Clang will not parse the type string correctly---unlike
-        // the old type parser---if there is no function name. The underscore
-        // itself isn't special, and will not end up being used as the function
-        // name in either case.
-        if (i == 0)
-            suffix = " _(";
-        else if (i == typeTokens.size() - 1)
-            suffix = ")";
-        else
-            suffix = ", ";
-
-        typeString += part + suffix;
-    }
-    typeString += ";";
-
-    std::string errors;
-    TypeParserResult tpResult;
-    auto ok = bv->ParseTypesFromSource(typeString, {}, {}, tpResult, errors);
-    if (ok && !tpResult.functions.empty()) {
-        auto functionType = tpResult.functions[0].type;
+	Ref<Type> retType;
+
+	std::vector<BinaryNinja::FunctionParameter> params;
+	bool varArg = false;
+	auto cc = bv->GetDefaultPlatform()->GetDefaultCallingConvention();
+
+	for (size_t i = 0; i < typeTokens.size(); ++i) {
+		Ref<Type> type = typeTokens[i].type.has_value() ? typeTokens[i].type.value() : Type::NamedType(bv, typeTokens[i].name);
+		if (i == 0)
+			retType = type;
+		else if (i == 1)
+		{
+			params.push_back({"self",
+				BinaryNinja::Type::NamedType(bv, ci.name),
+				true,
+				BinaryNinja::Variable()});
+		}
+		else
+		{
+			std::string argName;
+			if (i == 2)
+				argName = "sel";
+			else if (i - 3 < selectorTokens.size())
+				argName = selectorTokens[i - 3];
+
+			params.push_back({argName, typeTokens[i].type.has_value() ? typeTokens[i].type.value() : Type::NamedType(bv, typeTokens[i].name), true, BinaryNinja::Variable()});
+		}
+	}
+
+	auto functionType = BinaryNinja::Type::FunctionType(retType, cc, params, varArg);
+
+	// Search for the method's implementation function; apply the type if found.
+	if (auto f = bv->GetAnalysisFunction(bv->GetDefaultPlatform(), mi.implAddress))
+		f->SetUserType(functionType);
```

Comment:
Without looking too closely at what this is doing, my "need for a separate function" senses are going off. Seems to be a lot of indentation and complexity here which could benefit from either simplification or extraction into a helper function.